File: backend/src/models/Transaction.js
const mongoose = require('mongoose');

const transactionSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  amount: {
    type: Number,
    required: true,
    min: 0,
  },
  description: {
    type: String,
    required: true,
    trim: true,
  },
  type: {
    type: String,
    enum: ['income', 'expense'],
    required: true,
  },
  category: {
    type: String,
    enum: ['food', 'transportation', 'housing', 'utilities', 'entertainment', 'healthcare', 'other'],
    required: function() {
      return this.type === 'expense';
    },
  },
  date: {
    type: Date,
    default: Date.now,
  },
}, {
  timestamps: true,
});

// Middleware to update user totals when a transaction is created
transactionSchema.post('save', async function() {
  const User = mongoose.model('User');
  const user = await User.findById(this.userId);
  if (user) {
    await user.updateTotals(this.amount, this.type);
  }
});

// Middleware to update user totals when a transaction is deleted
transactionSchema.post('remove', async function() {
  const User = mongoose.model('User');
  const user = await User.findById(this.userId);
  if (user) {
    // Subtract the amount since we're removing the transaction
    await user.updateTotals(-this.amount, this.type);
  }
});

const Transaction = mongoose.model('Transaction', transactionSchema);

module.exports = Transaction; 

File: lib/models/transaction.dart
import 'package:json_annotation/json_annotation.dart';

part 'transaction.g.dart';

enum TransactionType {
  @JsonValue('income')
  income,
  @JsonValue('expense')
  expense
}

enum ExpenseCategory {
  @JsonValue('food')
  food,
  @JsonValue('transportation')
  transportation,
  @JsonValue('housing')
  housing,
  @JsonValue('utilities')
  utilities,
  @JsonValue('entertainment')
  entertainment,
  @JsonValue('healthcare')
  healthcare,
  @JsonValue('other')
  other
}

@JsonSerializable()
class Transaction {
  final String id;
  final String userId;
  final double amount;
  final String description;
  final DateTime date;
  final TransactionType type;
  final ExpenseCategory? category;

  Transaction({
    required this.id,
    required this.userId,
    required this.amount,
    required this.description,
    required this.date,
    required this.type,
    this.category,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) => _$TransactionFromJson(json);
  Map<String, dynamic> toJson() => _$TransactionToJson(this);
} 